src/edu/washington/escience/myria/operator/SampleWR.java

```diff
@@ -0,0 +1,110 @@
+package edu.washington.escience.myria.operator;
+
+import java.util.Arrays;
+import java.util.Random;
+
+import com.google.common.collect.ImmutableMap;
+
+import edu.washington.escience.myria.storage.TupleBatch;
+import edu.washington.escience.myria.storage.TupleBatchBuffer;
+
+public class SampleWR extends Sample {
+  /** Required for Java serialization. */
+  private static final long serialVersionUID = 1L;
+
+  /** True if operator has extracted sampling info. */
+  private boolean computedSamplingInfo = false;
+
+  /** Buffer for tuples that will be returned. */
+  private TupleBatchBuffer ans;
+
+  /** Current TupleID being processed. */
+  private int tupleNum = 0;
+
+  /** List of of indices that will be taken as samples. */
+  private int[] samples;
+
+  /** Current index of the samples array. */
+  private int curSampIdx = 0;
+
+  public SampleWR(final Operator left, final Operator right) {
+    super(left, right);
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws Exception {
+    // Extract sampling info from left operator.
+    if (!computedSamplingInfo) {
+      TupleBatch tb = getLeft().nextReady();
+      if (tb == null)
+        return null;
+      extractSamplingInfo(tb);
+
+      // Generate target indices to accept as samples.
+      samples = generateTargetSampleIndices(populationSize, sampleSize);
+
+      getLeft().close();
+      computedSamplingInfo = true;
+    }
+
+    // Return a ready tuple batch if possible.
+    TupleBatch nexttb = ans.popAny();
+    if (nexttb != null) {
+      return nexttb;
+    }
+    // Check if there's nothing left to sample.
+    if (curSampIdx >= samples.length) {
+      getRight().close();
+      setEOS();
+      return null;
+    }
+    Operator right = getRight();
+    for (TupleBatch tb = right.nextReady(); tb != null; tb = right.nextReady()) {
+      if (samples[curSampIdx] > tupleNum + tb.numTuples()) {
+        // nextIndex is not in this batch. Continue with next batch.
+        tupleNum += tb.numTuples();
+        continue;
+      }
+      while (curSampIdx < samples.length
+          && samples[curSampIdx] <= tupleNum + tb.numTuples()) {
+        ans.put(tb, samples[curSampIdx] - tupleNum);
+        curSampIdx++;
+      }
+      tupleNum += tb.numTuples();
+      if (ans.hasFilledTB()) {
+        return ans.popFilled();
+      }
+    }
+    return ans.popAny();
+  }
+
+  /**
+   * Generates a sorted sequence of random indices that will be chosen to sample
+   * from.
+   * 
+   * @param populationSize
+   *          size of the population that will be sampled from.
+   * @param sampleSize
+   *          number of samples to draw from the population.
+   * @return a sorted array of indices.
+   */
+  private int[] generateTargetSampleIndices(int populationSize, int sampleSize) {
```

Comment:
Sorry I wasn't clear, I meant logically parallel, in a single thread. I.e., for sample size `k`, you'd have `k` separate reservoirs of 1 element each, and you would evaluate the 1-element reservoir sampling algorithm independently `k` times for each tuple. Granted, having to generate `k` random numbers for each tuple might be a lot of overhead for large `k`...